### PR TITLE
refactor(*): Improve error tracking

### PR DESCRIPTION
**Pull request by grafakus**
### ✨ Description

**Related issue(s):** resolves https://github.com/grafana/explore-profiles/issues/263

This PR is an attempt to improve how we track errors, in order to benefit from great frontend observability dashboards and fast response in case of incidents!

### 📖 Summary of the changes

Besides adding an `ErrorBoundary` with a new error page:

image

This PR introduces several changes:
- A new `Logger` class, which encapsulates calls to Faro and/or `console.xyz()`
- All existing `console.xyz()` calls have been changed to calls to `logger.xyz()`
- Enforcing logger usage: ESLint will fail if any `console.xyz()` calls are present in the code
- When calling `displayError()`, the error is automatically logged
- When passing an `error` as prop to `<InlineBanner />`, it's automatically logged
- Refactor: `<InlineBanner />` now only accepts a single error as prop (no more arrays)

### 🧪 How to test?

The build should pass: some unit tests have been added, some refactored.

Manually:
- checkout this PR's branch locally,
- uncomment [these lines in faro-environments.ts](https://github.com/grafana/explore-profiles/pull/264/files#diff-920c144b32a862981a964657432fec328bff344602b837ee6678f30349ee0281R6-R13)
- manually force errors by modifying the code and playing with the app
- go to our "Frontend Observability" dashboard and see the errors and their the associated data (error context, ...)

### CODE REVIEW

File: src/app/components/Routes/Routes.tsx [hunk 1]
```diff
@@ -13,9 +13,6 @@ export function Routes() {
 
   return (
     <Switch>
-      <Route path={prefixRouteWithPluginBaseUrl(ROUTES.PROFILES_EXPLORER_VIEW)} exact>
```
Comment (grafakus) [inline src/app/components/Routes/Routes.tsx:16]: 🤨 🤷🏾‍♂️

File: src/shared/infrastructure/tracking/faro/faro.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import { GIT_COMMIT } from '../../../../version';
+import { getFaroEnvironment } from './getFaroEnvironment';
+
+let faro: Faro | undefined;
+
+export function getFaro() {
+  return faro;
+}
+
+export function initFaro() {
```
Comment (grafakus) [inline src/shared/infrastructure/tracking/faro/faro.ts:15]: TODO: prevent possibility to initialize it more than once?

Comment (grafakus) [reply src/shared/infrastructure/tracking/faro/faro.ts:14]: Done

File: src/shared/infrastructure/tracking/faro/faro.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  const { environment, faroUrl, appName } = faroEnvironment;
+
+  const { apps, bootData } = config;
+  const appRelease = apps[PYROSCOPE_APP_ID].version;
+  const userEmail = bootData.user.email;
+
+  setFaro(
+    initializeFaro({
```
Comment (grafakus) [inline src/shared/infrastructure/tracking/faro/faro.ts:31]: See https://grafana.com/docs/grafana-cloud/monitor-applications/frontend-observability/instrument/faro/

File: src/shared/infrastructure/tracking/faro/faro.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        email: userEmail,
+      },
+      instrumentations: [
+        ...getWebInstrumentations({
+          captureConsole: false,
+        }),
+        new TracingInstrumentation(),
+      ],
+      isolate: true,
+      beforeSend: (event) => {
```
Comment (ifrost) [inline src/shared/infrastructure/tracking/faro/faro.ts:49]: BTW. just ouf of curiosity - Is this needed? Isn't Faro backend/app configured to accept logs from specific routes? 🤔  What is preventing anyone else from pushing to the same collector some rubbish data? 🤔

File: src/app/App.tsx [hunk 5]
```diff
@@ -1,30 +1,42 @@
...
+  }
 
   return (
-    <QueryClientProvider client={queryClient}>
-      <Onboarding>
-        <div className={styles.pageContainer}>
-          <PluginPage layout={PageLayoutType.Custom}>
-            <div className="pyroscope-app">
-              <Routes />
+    <ErrorBoundary onError={setError}>
```
Comment (ifrost) [inline src/app/App.tsx:25]: `ErrorBoundary` from `@grafana/ui` pushes to faro already in  https://github.com/grafana/grafana/blob/47f87171496f675bc44ddf839effaed1846bfa4e/packages/grafana-ui/src/components/ErrorBoundary/ErrorBoundary.tsx#L38. But how does it work when faro is initialized twice - in grafana and in the app? Is the same error reported twice?

**Review by ifrost — APPROVED** [review]

**Pull request merged**